File: src/Polyphonic.TelegramBot/Infrastructure/BotUpdateHandler.cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Polyphonic.TelegramBot.Abstractions;
using Polyphonic.TelegramBot.Helpers;
using Telegram.Bot;
using Telegram.Bot.Polling;
using Telegram.Bot.Types;
using Telegram.Bot.Types.Enums;

namespace Polyphonic.TelegramBot.Infrastructure;

internal class BotUpdateHandler(
    IServiceProvider serviceProvider,
    ILogger<BotUpdateHandler> logger) : IUpdateHandler
{
    public Task HandlePollingErrorAsync(
        ITelegramBotClient botClient,
        Exception exception,
        CancellationToken cancellationToken)
    {
        logger.LogError(exception, "Exception happened during bot API polling");

        return Task.CompletedTask;
    }

    public async Task HandleUpdateAsync(
        ITelegramBotClient botClient,
        Update update,
        CancellationToken cancellationToken)
    {
        switch (update.Type)
        {
            // A message was received
            case UpdateType.Message:
                await HandleMessage(botClient, update.Message!, cancellationToken);
                break;

            // A button was pressed
            case UpdateType.CallbackQuery:
                await HandleButtonClick(botClient, update.CallbackQuery!, cancellationToken);
                break;
            
            // An inline query was received
            case UpdateType.InlineQuery:
            { 
                // braces to make using declaration work
                
                using var serviceScope = serviceProvider.CreateScope();
                var inlineQueryHandler = serviceScope.ServiceProvider.GetRequiredService<IBotInlineQueryHandler>();
                
                await inlineQueryHandler.HandleAsync(botClient, update.InlineQuery!, cancellationToken);
                break;
            }
        }
    }

    private async Task HandleMessage(
        ITelegramBotClient botClient,
        Message message,
        CancellationToken cancellationToken)
    {
        var (hasSender, sender) = message.GetSender();
        var (hasMessageText, messageText) = message.GetMessageText();

        if (!hasSender)
        {
            logger.LogError("Can't handle message. User Id is null");
            
            return;
        }

        if (!hasMessageText)
        {
            logger.LogError("Can't handle message. Message text is null");

            return;
        }
        
        logger.LogInformation("{UserName} wrote {MessageText}", sender.FirstName, messageText);
        
        if (IBotCommandHandler.IsCommand(messageText))
        {
            // means message is bot command
            await HandleCommand(botClient, message, cancellationToken);
        }
        else
        {
            // this is equivalent to forwarding, without the sender's name
            
            await botClient.CopyMessageAsync(
                sender.Id,
                sender.Id,
                message.MessageId,
                cancellationToken: cancellationToken);
        }
    }

    private async Task HandleCommand(
        ITelegramBotClient botClient,
        Message message,
        CancellationToken cancellationToken)
    {
        var parsedCommand = IBotCommandHandler.ParseCommand(message.Text);

        if (!parsedCommand.IsCommandValid)
        {
            await botClient.SendTextMessageAsync(
                message.From!.Id,
                $"Invalid command '{parsedCommand.CommandName}'. Command format : '/<command_name> [argument_1] ... [argument_n]'",
                cancellationToken: cancellationToken);
        }

        using var serviceScope = serviceProvider.CreateScope();

        var commandHandlers = 
            serviceScope.ServiceProvider.GetServices<IBotCommandHandler>();

        foreach (var commandHandler in commandHandlers)
        {
            if (!commandHandler.CanHandle(parsedCommand).CanHandleInMessage)
            {
                continue;
            }

            await commandHandler.HandleAsync(botClient, message, parsedCommand, cancellationToken);
            
            return;
        }
    }

    private async Task HandleButtonClick(
        ITelegramBotClient botClient,
        CallbackQuery query,
        CancellationToken cancellationToken)
    {
        throw new NotImplementedException();
        
        string text = string.Empty;
        //InlineKeyboardMarkup markup = new(Array.Empty<InlineKeyboardButton>());

        // if (query.Data == GET_SONG_LINK_BUTTON)
        // {
        //     text = GETTING_SONG_LINK_MESSAGE;
        //     //markup = _printMenuMarkup;
        // }

        // Close the query to end the client-side loading animation
        await botClient.AnswerCallbackQueryAsync(query.Id, cancellationToken: cancellationToken);

        // Replace menu text and keyboard
        await botClient.EditMessageTextAsync(
            query.Message!.Chat.Id,
            query.Message.MessageId,
            text,
            ParseMode.Html,
            //replyMarkup: markup,
            cancellationToken: cancellationToken);
    }
}

File: src/Polyphonic.TelegramBot/Infrastructure/BotUpdateHandler.fs
namespace Polyphonic.TelegramBot.Infrastructure

open System
open System.Threading.Tasks
open Microsoft.Extensions.Logging
open Telegram.Bot
open Telegram.Bot.Types
open Telegram.Bot.Types.Enums
open Telegram.Bot.Types.ReplyMarkups
open Telegram.Bot.Polling

type BotUpdateHandler(logger: ILogger<BotUpdateHandler>) =
    let mutable isScreaming = false

    let firstMenu = "<b>Menu 1</b>\n\nA beautiful menu with a shiny inline button."

    let secondMenu =
        "<b>Menu 2</b>\n\nA better menu with even more shiny inline buttons."

    // Values that are intended to be constants can be marked with the Literal attribute.
    // This attribute has the effect of causing a value to be compiled as a constant.
    [<Literal>]
    let nextButtonText = "Next"

    [<Literal>]
    let backButtonText = "Back"

    [<Literal>]
    let tutorialButtonText = "Tutorial"

    let firstMenuMarkup =
        InlineKeyboardMarkup(InlineKeyboardButton.WithCallbackData(nextButtonText))

    let backButton = [| InlineKeyboardButton.WithCallbackData(backButtonText) |]

    let tutorialButton =
        [| InlineKeyboardButton.WithUrl(tutorialButtonText, "https://core.telegram.org/bots/tutorial") |]

    let secondMenuMarkup =
        InlineKeyboardMarkup(inlineKeyboard = [| backButton; tutorialButton |])

    let sendMenu (botClient: ITelegramBotClient) (userId: int64) =
        async {
            botClient.SendTextMessageAsync(userId, firstMenu, parseMode = ParseMode.Html, replyMarkup = firstMenuMarkup)
            |> Async.AwaitTask
            |> ignore
        }
        |> Async.StartAsTask

    let handleCommand (botClient: ITelegramBotClient, userId: int64, command: string) =
        async {
            match command with
            | "/scream" -> isScreaming <- true
            | "/whisper" -> isScreaming <- false
            | "/menu" -> do! sendMenu botClient userId |> Async.AwaitTask
            | _ -> ()
        }
        |> Async.StartAsTask

    let handleButton (botClient: ITelegramBotClient, query: CallbackQuery) =
        async {
            let mutable text = ""
            let mutable markup = new InlineKeyboardMarkup(inlineKeyboard = [||])

            match query.Data with
            | "Next" ->
                text <- secondMenu
                markup <- secondMenuMarkup
            | "Back" ->
                text <- firstMenu
                markup <- firstMenuMarkup
            | _ -> ()

            do! botClient.AnswerCallbackQueryAsync(query.Id) |> Async.AwaitTask

            botClient.EditMessageTextAsync(
                chatId = query.Message.Chat.Id,
                messageId = query.Message.MessageId,
                text = text,
                parseMode = ParseMode.Html,
                replyMarkup = markup
            )
            |> Async.AwaitTask
            |> ignore
        }
        |> Async.StartAsTask

    interface IUpdateHandler with
        member this.HandlePollingErrorAsync(botClient, e: Exception, cancellationToken) =
            logger.LogError(e.Message)
            Task.CompletedTask

        member this.HandleUpdateAsync(botClient, update, cancellationToken) =
            task {
                match update.Type with
                | UpdateType.Message ->
                    let message = update.Message
                    let user = message.From
                    let text: string = message.Text

                    let username = user.FirstName
                    Console.WriteLine($"{username} wrote {text}")

                    if text.StartsWith("/") then
                        do! handleCommand (botClient, user.Id, text) |> Async.AwaitTask
                    elif isScreaming && text.Length > 0 then
                        botClient.SendTextMessageAsync(user.Id, text.ToUpper(), entities = message.Entities)
                        |> Async.AwaitTask
                        |> ignore
                    else
                        let! _ =
                            botClient.CopyMessageAsync(user.Id, user.Id, message.MessageId)
                            |> Async.AwaitTask

                        () // same as |> ignore ???

                    ()
                | UpdateType.CallbackQuery ->
                    do!
                        handleButton (botClient, update.CallbackQuery)
                        |> Async.AwaitTask
                        |> Async.Ignore
                | _ -> ()
            }
